python/cudf/cudf/core/udf/utils.py

```diff
@@ -41,6 +41,9 @@
 from cudf.utils.performance_tracking import _performance_tracking
 from cudf.utils.utils import initfunc
 
+if TYPE_CHECKING:
+    from collections.abc import Callable
```

Comment:
I see an instance below with `Callable` being used as a type annotation without quotes. Wouldn't it fail at runtime in that case due to the `TYPE_CHECKING` guard?

Reply:
That's what I would have thought too. For example:

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Callable

class StuffDoer:
    def do_stuff(self, f: Callable) -> None:
        print(f.__name__)

def somefunc():
    pass

StuffDoer().do_stuff(somefunc)
```

fails like this:

> NameError: name 'Callable' is not defined. Did you mean: 'callable'?

... but there are other hints in these files which refer unquoted and unconditionally to things imported in a `TYPE_CHECKING` block.

For example:

https://github.com/rapidsai/cudf/blob/1a2aad27b7e136f87be80debed6da7d3528ebda1/python/cudf/cudf/core/column/numerical.py#L30-L32

https://github.com/rapidsai/cudf/blob/1a2aad27b7e136f87be80debed6da7d3528ebda1/python/cudf/cudf/core/column/numerical.py#L195

When I first saw that, I thought "ok something is going on here that I don't understand, but might as well be consistent with the rest of the code" around these changes.

I did some research thanks to your review comment and figured out what's going on .... the use of `from __future__ import annotations` defers evaluation of type hints.

* https://peps.python.org/pep-0563/
* https://mypy.readthedocs.io/en/stable/runtime_troubles.html#future-annotations-import-pep-563

Adding a `from __future__ import annotations` to the top of that example code allows it to run without issue.

Summary from those `mypy` docs:

> *Many of the issues described here are caused by Python trying to evaluate annotations. Future Python versions (potentially Python 3.12) will by default no longer attempt to evaluate function and variable annotations. This behaviour is made available in Python 3.7 and later through the use of from `__future__ import annotations`.*

So I guess here in `cudf`, the type hints like these really should be left unquoted (for consistency with the rest of the surrounding code), in favor of using that `from __future__ import annotations` mechanism.

Reply:
> So I guess here in cudf, the type hints like these really should be left unquoted (for consistency with the rest of the surrounding code), in favor of using that from __future__ import annotations mechanism.

Agree, however I suggest only using `if TYPE_CHECKING:` to avoid circular imports. No need to muddle up the import section.

Reply:
> Agree, however I suggest only using if TYPE_CHECKING: to avoid circular imports.

We would need to ignore TCH003 for that.

Reply:
Wow [TCH003](https://docs.astral.sh/ruff/rules/typing-only-standard-library-import/) sounds like a regression to me: clutter the import section just to replace the overhead of a (cached) import statement with a if statement.

Anyways, if that is what `ruff` wants, I guess it is fine :)